### PR TITLE
[approvaltest] Ignore dynamic fields from sort fields

### PR DESCRIPTION
**Pull request by lahsivjar**
The approval tests doesn't handle dynamic fields correctly, the fields are sorted before `dynamic` is handled. Due to this, 2 issues could happen:

1. If a field is passed as `dynamic` but it is also in the sort fields then the resulting sort would be indeterministic.
2. If all the sort fields are identical then the fields are compared using `bytes.Compare` this will again result in indeterministic sorts if the dynamic fields are randomly generated.

Both these issues are fixed in this PR by rewriting the dynamic fields before sorting and comparing for approvals.

Required to fix the failures in https://github.com/elastic/apm-server/pull/13196

### CODE REVIEW

**Review by axw — COMMENTED** [review]
Could this PR be simplified by pulling the rewrite loops out of approveEventDocs and approveFields? i.e. change ApproveEvents and ApproveFields to first rewrite, then sort, and finally approve.

**Comment by lahsivjar** [comment]
> Could this PR be simplified by pulling the rewrite loops out of approveEventDocs and approveFields?

@axw This was my first thought and I spent some time into trying to refactor code as per this but the code I came up with was not optimum and required copying the raw fields and the raw source both in-memory. My previous effort was based on not decoding the data first but I think we might be able to do this if we decode to JSON first and then parse dynamic fields. I will give this a shot.

**Comment by lahsivjar** [comment]
@axw Fixed the logic to rewrite before sort. PTAL.

File: pkg/approvaltest/approvals.go [hunk 1]
```diff
@@ -115,22 +111,42 @@ func approveEventDocs(t testing.TB, name string, eventDocs [][]byte, dynamic ...
...
 	}, dynamic...)
 
-	// Rewrite all dynamic fields to have a known value,
-	// so dynamic fields don't affect diffs.
-	events := make([]interface{}, len(eventDocs))
-	for i, doc := range eventDocs {
+	for i := range srcs {
 		for _, field := range dynamic {
-			existing := gjson.GetBytes(doc, field)
+			field = strings.ReplaceAll(field, ".", "\\.")
```
Comment (axw) [inline pkg/approvaltest/approvals.go:116]: Is this going to work for `_source` as well as `fields`? We might need to switch based on whether it's ApproveEvents vs. ApproveFields

**Review by axw — COMMENTED** [review]
Looks good, just one query

File: pkg/approvaltest/approvals.go (hunk 1 above)
Comment (lahsivjar) [reply pkg/approvaltest/approvals.go]: @axw You are right, it wouldn't have worked with `ApproveEvents`, I have fixed it now. Thanks for catching this.

**Review by axw — APPROVED** [review]

**Pull request merged**